actionpack/lib/action_controller/metal/mime_responds.rb

```diff
@@ -209,7 +209,7 @@ def respond_to(*mimes)
           raise ActionController::RespondToMismatchError
         end
         _process_format(format)
-        _set_rendered_content_type format
+        _set_rendered_content_type(format) unless collector.any_response?
```

Comment:
`_set_rendered_content_type` has a guard against setting the `Content-Type`/`media_type`:
https://github.com/rails/rails/blob/4f9f9ca23a39b2b529782f957be645741a20cf20/actionpack/lib/action_controller/metal/rendering.rb#L74-L76

and `AbstractController::Rendering#render` uses that method passing the `rendered_format`:
https://github.com/rails/rails/blob/4f9f9ca23a39b2b529782f957be645741a20cf20/actionpack/lib/abstract_controller/rendering.rb#L29

which is `text/plain` by default in `AbstractController::Rendering`:
https://github.com/rails/rails/blob/4f9f9ca23a39b2b529782f957be645741a20cf20/actionpack/lib/abstract_controller/rendering.rb#L55-L56

but is overridden by `ActionView::Rendering`:
https://github.com/rails/rails/blob/4f9f9ca23a39b2b529782f957be645741a20cf20/actionview/lib/action_view/rendering.rb#L29

In the case of `render json:` (and similarly `:js` and `:xml`), the response content type is set during the call to `render`, again with a guard against overriding any existing value, because you could have set it manually before `render` or with `render(content_type: :csv)`.
https://github.com/rails/rails/blob/4f9f9ca23a39b2b529782f957be645741a20cf20/actionpack/lib/action_controller/metal/renderers.rb#L166

---

All of this to say that I also tried moving the call to `_set_rendered_content_type` after the call to `response.call` without the guard that you introduced, and it also resulted in the right `Content-Type` being set (the one from rendering, not the one from the request `Accept` header).

I think that would make more sense as the response content type should come more from what happens during rendering than what has been asked. e.g. what should we return when we do this? 

```ruby
      format.xml { render json: { foo: "bar" } }
```

IMO JSON, but that would take a pass of deprecation warnings. The problem doing it this way is that now we have a compatibility issue because it impacts more than the `any` case (which I agree was fine to change). Also doing this makes a lot of test fail and I haven't yet figured out why so we probably need to store the content type that was asked and use it unless another one was set during rendering.

Reply:
Thanks for digging in this rabbit hole with me :) 

> All of this to say that I also tried moving the call to _set_rendered_content_type after the call to response.call

Originally I went that path and thought we should modify the `respond_to` mehod entirely, because as you mentioned in your example `format.xml { render json: { foo: "bar" } }` returning a `text/xml` response type seemed wrong.

However after second reflexion I think returning `text/xml` makes sense. Let me clarify why.
If a controller action specify how to handle a specific format `xml` in that case, the server knows that the client expects (and in some certain cases, only accepts) a XML response, returning a JSON response breaks the contract and is a developer mistake. 

On the other hand, it's impossible for an application to handle all formats inside `format.any` and thus the framework shouldn't make assumption that the server will for sure return a response based on what the client asked.